File: backend/controllers/report_user_controllers.go
package controllers

import (
	"context"
	"fmt"
	"time"

	"eunity.com/backend-main/helpers/DBManager"
	"eunity.com/backend-main/models"
	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Report_User_controllers struct{}

// @Summary Add Report User
// @Schemes
// @Description Adds a new User Report to the database
// @Tags Report User
// @Accept mpfd
// @Produce json
// @Param reported_user formData string true "Reported User"
// @Param rule_violations formData []string true "Rule Violations"
// @Param report_comments formData string false "Additional comments that explain why the user was reported"
// @Success 200 {string} Successfully Added Report
// @Success 400 {string} Couldn't Add Report
// @Router /report_user/add_report [post]
func (ru *Report_User_controllers) POST_Add_Report(c *gin.Context) {
	//Get the user ID from the cookie so we know who is requesting to find their open report
	user_id := c.Keys["user_id"].(string)

	//Bind the params to a UserReport
	var report models.UserReport
	err := c.Bind(&report)
	if err != nil {
		fmt.Println(err)
		c.JSON(400, gin.H{"response": "Error in Binding JSON"})
		return
	}

	//Init a new Object ID
	objectID := primitive.NewObjectID()

	//Verify Rule Violations are present
	if report.RuleViolations == nil || len(report.RuleViolations) <= 0 {
		c.JSON(400, gin.H{"response": "Missing Rule Violations"})
		return
	}

	//Verify Reported User is specified
	if report.ReportedUser == "" {
		c.JSON(400, gin.H{"response": "Missing Reported User"})
		return
	}

	//Init a new UserReport struct
	new_report := models.UserReport{
		ID:             &objectID,
		ReportedUser:   report.ReportedUser,
		ReportedBy:     user_id,
		RuleViolations: report.RuleViolations,
		ReportComments: report.ReportComments,
		ReportedAt:     time.Now(),
	}

	//Attempt to insert into db
	_, err = DBManager.DB.Collection("user_reports").InsertOne(context.Background(), new_report)
	if err != nil {
		c.JSON(400, gin.H{
			"response": "Unable to Insert Report",
		})
		return
	}

	//Return success code
	c.JSON(200, gin.H{
		"response": "Successfully Created Report",
	})
}

File: python/controllers/report_user_controllers.py
from fastapi import FastAPI, Form, Response
from pydantic import BaseModel
from typing import List
import datetime
from bson import ObjectId
from pymongo import MongoClient
from fastapi import APIRouter
from models.user_report import UserReport





router = APIRouter()

@router.post("/report_user/add_report")
async def post_add_report(self, reported_user: str = Form(...), rule_violations: List[str] = Form(...), report_comments: str = Form(None), user_id: str = None):
    if user_id is None:
        return Response(status_code=400, content="User ID is required")

    report_data = UserReport(reported_user=reported_user, rule_violations=rule_violations, report_comments=report_comments)

    if not report_data.rule_violations:
        return Response(status_code=400, content="Rule Violations are required")
    if not report_data.reported_user:
        return Response(status_code=400, content="Reported User is required")

    new_report = {
        "_id": ObjectId(),
        "reported_user": report_data.reported_user,
        "reported_by": user_id,
        "rule_violations": report_data.rule_violations,
        "report_comments": report_data.report_comments,
        "reported_at": datetime.datetime.utcnow(),
    }

    try:
        self.db["user_reports"].insert_one(new_report)
        return Response(status_code=200, content="Successfully Created Report")
    except:
        return Response(status_code=400, content="Unable to Insert Report")

